Update C++ rules (#3220)
SonarSource/rspec: rules/S5274/cfamily/rule.adoc

== Why is this an issue?

> Usually, when copying an object, the source object is unchanged, meaning all resources owned by the source objects must be duplicated during the copy operation. If the source object is no longer used, this duplication is inefficient. Since {cpp}11, a move semantic mechanism has been added to detect such cases and replace the expensive copy with a much cheaper move operation that will steal resources.

Reviewer: re steal v transfer.
```suggestion
Usually, when copying an object, the source object is unchanged, meaning all resources owned by the source objects must be duplicated during the copy operation. If the source object is no longer used, this duplication is inefficient. Since {cpp}11, a move semantic mechanism has been added to detect such cases and replace the expensive copy with a much cheaper move operation that will transfer resources.
```

The cornerstone of move semantics is detecting during a "copy" whether the source object will be reused or not. There are three situations:

* The object is a temporary object with no name, and if it can't be named, it can't be used
* The object is used in some specific places, such as a return statement

> * The user explicitly promises to the compiler that he won't care for the object's current value any longer. He does so by using the specific cast operation named ``++std::move++``.

Reviewer: Make it gender neutral with "they":
```suggestion
* The user explicitly promises to the compiler that they won't care for the object's current value any longer. They do so by using the specific cast operation named ``++std::move++``.
```

[See the second definition in the Oxford dictionary](https://www.oxfordlearnersdictionaries.com/definition/english/they)

Author: Indeed, you're right, thanks!

If the user writes ``++std::move++`` in one situation that is already handled by the first two cases, it has two drawbacks:

* It is clumsy, useless code, which makes understanding the code more complex
* In some cases, it can decrease performances because this can deactivate another optimization of the compiler, named copy elision.

When copy elision occurs, the object is neither copied nor moved (even if the copy/move constructors have side effects). The two objects are collapsed into only one memory location. The moment when copy elision occurs is compiler-dependent but is mandatory in the following cases:

* in a return statement, if the returned object is a prvalue of the same class type as the function return type
* in the initialization of a variable, if the initializer expression is a prvalue of the same class type as the variable type

This rule reports an issue when the use of ``++std::move++`` prevents the copy elision from happening.

=== Noncompliant code example

> [source,cpp,diff-id=1,diff-type=noncompliant]

Reviewer: In the example below, we have two comments "Noncompliant, prevents copy elision", and one comment "Noncompliant". Given the sentence above, the comments could all be simply "Noncompliant". WDYT?

Author: Agreed, fixed in https://github.com/SonarSource/rspec/pull/3220/commits/a35d25859fc2556c45fdf8d05af34d7662897315

Reviewer: You kept the first one. I'm not sure it's intentional. I don't mind either way. Feel free to mark this discussion as resolved.

Author: Yes, it was intentional.
For example, for the user who starts by reading the code before the description :)

> ----

Reviewer: `vector` is the only thing that is not fully qualified. I suggest writing `std::vector` instead.

Author: Good catch, fixed in https://github.com/SonarSource/rspec/pull/3220/commits/a35d25859fc2556c45fdf8d05af34d7662897315
